test/meta_testing_tools.jl

```diff
@@ -71,14 +71,38 @@ function fails(f)
         did_fail |= result isa Test.Fail
         if result isa Test.Error
             # Log a error message, with original backtrace
-            show(result)
             # Sadly we can't throw the original exception as it is only stored as a String
             error("Error occurred during `fails`")
         end
     end
     return did_fail
 end
 
+"""
+    errors(f, msg_pattern="")
+
+`errors(f, msg_pattern)` returns true if at least 1 error is recorded into a testset,
+    with a failure matching the given pattern.
+
+`f` should be a function that takes no argument, and calls some code that uses `@testset`.
+`msg_pattern` is a regex or a string, that should be contained in the error message.
+If nothing is passed then it default to the empty string, which matches any error message.
+
+If a test fails (rather than passing or erroring) then `errors` will throw an error.
+"""
+function errors(f, msg_pattern="")
+    results = nonpassing_results(f)
+
+    for result in results
+        result isa Test.Fail && error("Test actually failed (nor errored): \n $result")
+        result isa Test.Error && occursin(msg_pattern, result.value) && return true
+    end
+    return false  # no matching error occured
+end
+
+
```

Comment:
maybe overload `Base.show` to not print the `results`? 

```suggestion
Base.show(::IO, ts::NonPassingTestset) = print(typeof(ts), "(\"", ts.description, "\")")
```

I know it's meant for programmatic useage, but it leads to slightly confusing REPL output (stemming from the default pretty-printing of `Test.Result`s)

e.g. currently

```julia
julia> ts = @testset NonPassingTestset "foo" begin
           @test false
           @test true
           @test error()
       end
NonPassingTestset("foo", Any[Test Failed at REPL[23]:2
  Expression: false, Test Passed, Error During Test at REPL[23]:4
  Test threw exception
  Expression: error()

  Stacktrace:
   [1] error() at ./error.jl:42
   [2] top-level scope at REPL[23]:4
   [3] top-level scope at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115
   [4] top-level scope at REPL[23]:2
  ])
```
whereas it might be nicer to just have
```julia
julia> ts = @testset NonPassingTestset "foo" begin
           @test false
           @test true
           @test error()
       end
NonPassingTestset("foo")
```

Reply:
While that output is confusing, it was useful to me during debugging.
Helped me spot a mistake i made; becuase i wasn't pusing contents up correctly.
Your suggested output would hide that info,
WHile there is a more useful output that is cleaner, working it out would be effort.

Reply:
yah, i think a better output would be something along the lines of `NonPassingTestset("foo", Any[Fail, Pass, Error])` but i'm not sure the code is worth it.